File: Day-12/solution.py
def is_anagram(s1, s2):
    # Step 1: Length check
    if len(s1) != len(s2):
        return False

    freq = {}

    # Step 2: Count frequency of characters in s1
    for ch in s1:
        if ch in freq:
            freq[ch] += 1
        else:
            freq[ch] = 1

    # Step 3: Subtract frequency using s2
    for ch in s2:
        if ch not in freq:
            return False
        freq[ch] -= 1
        if freq[ch] < 0:
            return False

    return True


# Testing
print(is_anagram("listen", "silent"))  # True
print(is_anagram("hello", "world"))    # False

File: Day-12/solution.cpp
#include <iostream>
#include <unordered_map>
using namespace std;

bool isAnagram(string s1, string s2) {
    if (s1.length() != s2.length())
        return false;

    unordered_map<char, int> freq;

    // Count characters of first string
    for (char c : s1) {
        freq[c]++;
    }

    // Subtract using second string
    for (char c : s2) {
        freq[c]--;
        if (freq[c] < 0)  // Early exit optimization
            return false;
    }

    return true;
}

int main() {
    cout << isAnagram("listen", "silent") << endl; // 1 (True)
    cout << isAnagram("hello", "world") << endl;   // 0 (False)
}
